src/Shared/Buffers/RefPooledArrayBufferWriter.cs

```diff
@@ -0,0 +1,266 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// Copied from https://github.com/dotnet/corefx/blob/b0751dcd4a419ba6731dcaa7d240a8a1946c934c/src/System.Text.Json/src/System/Text/Json/Serialization/ArrayBufferWriter.cs
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers;
+
+/// <summary>
+/// A high-performance struct-based IBufferWriter&lt;byte&gt; implementation that uses ArrayPool for allocations.
+/// Designed for zero-allocation scenarios when used with generic methods via `allows ref struct` constraint.
+/// </summary>
+internal ref struct RefPooledArrayBufferWriter<T> : IBufferWriter<T>, IDisposable
+{
+    private T[] _rentedBuffer;
+    private Span<T> _buffer;
+    private int _index;
+
+    private const int MinimumBufferSize = 256;
+
+    /// <summary>
+    /// Initializes the <see cref="RefPooledArrayBufferWriter{T}"/> with initial buffer. 
+    /// </summary>
+    /// <param name="initialBuffer">The initial buffer to start writer with.</param>
+    public RefPooledArrayBufferWriter(Span<T> initialBuffer)
+    {
+        // no rented buffer initially - only if we need to grow over the limits of initialBuffer
+        _rentedBuffer = null!;
+
+        _buffer = initialBuffer;
+        _index = 0;
```

Comment:
I don't think this comment is necessary once we properly mark the field as nullable.

```suggestion
        _buffer = initialBuffer;
```

Reply:
removed unnecessary code